src/core/gui/XournalppCursor.cpp

```diff
@@ -515,6 +521,71 @@ auto XournalppCursor::createHighlighterOrPenCursor(int size, double alpha) -> Gd
 }
 
 
+auto XournalppCursor::createHorizontalLineCursor() -> GdkCursor* {
+    GtkWidget* theWidget = control->getWindow()->getXournal()->getWidget();
+    GdkWindow* theWindow = gtk_widget_get_window(theWidget);
+
+    // Get the position of the mouse cursor within the window
+    gint x, y;
+    GdkDevice *mouse_device;
+
+    #if GTK_CHECK_VERSION (3,20,0)
+    GdkSeat *seat = gdk_display_get_default_seat (gdk_display_get_default ());
+    mouse_device = gdk_seat_get_pointer (seat);
+    #else
+    GdkDeviceManager *devman = gdk_display_get_device_manager (gdk_display_get_default ());
+    mouse_device = gdk_device_manager_get_client_pointer (devman);
+    #endif
+
+    gdk_window_get_device_position (theWindow, mouse_device, &x, &y, nullptr);
+
+    Layout* const layout = this->control->getWindow()->getLayout();
+    XojPageView* page = layout->getPageViewAt(x, y);
+    if (page) {
+        Range range = page->getVisiblePart();
+        double zoom = control->getZoomControl()->getZoom();
+        const auto width = static_cast<int32_t>(std::round(range.getWidth() * zoom));
+        const int height = 5; // Five is the default cursor size used elsewhere in the app.
+        auto dx = (int32_t) (x - static_cast<int32_t>(range.getX() * zoom) - page->getX());
+        // 64-bit field containing the current value of dx and width to check
+        // if we need to re-draw the cursor
+        auto flavour = (int64_t) (dx << (sizeof(gint) * 4) | width);
```

Comment:
I think this expression has a few mistakes:
* `dx` is int32_t, so `(dx << (sizeof(gint) * 4))` is of type `int32_t`. `dx` needs to be upcast to `int64_t` first to force the operations to be over 64 bit integers.
* We need `sizeof(decltype(width))` instead of `sizeof(gint)` to match the exact size of `width` in  bytes.
* Modern processors define 1 byte to be 8 bits, so this should be `* 8` and not `* 4`.
```suggestion
        auto flavour = static_cast<int64_t>(dx) << (sizeof(decltype(width)) * 8) | width;
```

Also:
* `this->currentCursorFlavour` is defined as `long`, which is only guaranteed a size of at least 32 bits. It may be possible for `long` to be defined as 32 bits, in which case `flavour` will be truncated. `this->currentCursorFlavour` should probably be refined as a long long.

Reply:
Ah, this was kind of why I wanted to avoid hashing it and just saving the value of DX.

I think we could probably re-evaluate the types of the involved variables. You're right, I also messed up the shift, and it's half what it should have been. I played around with a minimal code sample to confirm all this like I should have in the first place.

To the final point, I'd worry about re-defining `this->currentCursorFlavor`, since it's used so many other places with even more dubious bit hacking going on.

So, I think it's fine to explicity cast as an int64_t there, but if the precision of `this->currentCursorFlavor` is an issue, then we might need to re-think this hashing, to see if we can get it to fit in 32 bits while still having sufficiently unique values. All we need from it is, "have these two numbers changed?" A checksum, if you will :)

Reply:
I did some debugging to see if I can check out what the hash looks like, and i came across this:
```
(gdb) p sizeof(gulong)
$1 = 8
```
I know it's probably not guaranteed, but on my computer, a modern, 64-bit Intel-based ThinkPad, `gulong` is 8 bytes, 64 bits... That's enough, isn't it? I stepped through the process of constructing that number, and it looks like that'd work fine.

My current understanding of the process is this

First, cast dx to a `gulong`, a 64-bit unsigned integer.
```C++
auto flavour = static_cast<gulong>(dx);
```

Next, shift it over by 32 bits
```C++
flavour <<= (sizeof(decltype(width)) * 8);
```

Then, bitwise `or` that with the width value
```C++
flavour |= width;
```

And you have a hash that looks like this.
```
0b00000000_00000000_00000001_01110110_00000000_00000000_00000010_01010011
```
That gets saved as a `gulong` in `this->currentCursorFlavor`, and bob's your uncle.